### PR TITLE
#79: Added system requirements and UML model.

### LINKED ISSUE #47: Support GEOMETRY without SRID
## Situation

In the VS API towards the core database, the SRID parameter of the `GEOMETRY` type is mandatory. It is optional when creating columns through `CREATE TABLE` though.

This means there is a default value. I was not able to find it in the documentation yet.

## Acceptance Criteria

* Clarified default value
* Added support for `GEOMETRY` without SRID

### LINKED ISSUE #50: Improve running Lua checks
## Situation

Currently the static code analysis and coverage checks only run on Travis CI. This way you don't realize if they fail locally causing you to check in incomplete sources now and then.

## Acceptance criteria

* Coverage check runs with unit tests
* Static code analysis runs with unit tests

### LINKED ISSUE #79: Migrate system requirements and design
## Acceptance Criteria

1. System requirement migrated from `row-level-security` (Java).
2. Design migrated and adapted from `row-level-security`.
3. Requirement tracing green.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #79.
Closes #47.
Closes #50.

### CODE REVIEW

File: doc/model/diagrams/sequence/seq_entry.plantuml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+activate compiler
+compiler -> entry : adapter_call(request : JSON string)
+activate entry
+create rls_adapter
+entry -> rls_adapter : create
+create request_dispatcher
+entry -> request_dispatcher : create
+entry -> request_dispatcher : init(rls_adapter)
+activate request_dispatcher
+request_dispatcher -->> entry :this
```
Comment (AnastasiiaSergienko) [inline doc/model/diagrams/sequence/seq_entry.plantuml:20]: What does "this" mean?

File: src/main/lua/entry.lua [hunk 2]
```diff
@@ -0,0 +1,19 @@
+---
+-- This is the main entry point of the concrete Lua Virtual Schema adapter.
+-- <p>
+-- It is responsible creating for wiring up the adapter objects.
```
Comment (AnastasiiaSergienko) [inline src/main/lua/entry.lua:4]:
```suggestion
-- It is responsible for wiring up the adapter objects.
```

File: src/main/lua/entry.lua [hunk 3]
```diff
@@ -0,0 +1,19 @@
+---
+-- This is the main entry point of the concrete Lua Virtual Schema adapter.
```
Comment (AnastasiiaSergienko) [inline src/main/lua/entry.lua:2]:
```suggestion
-- This is the main entry point of the Lua Virtual Schema adapter.
```

File: src/main/lua/exasolvs/request_dispatcher.lua [hunk 4]
```diff
@@ -1,15 +1,42 @@
 local log = require("remotelog")
 local cjson = require("cjson")
-local adapter = require("exasolrls.adapter", "adapter")
+--local adapter = require("exasolrls.adapter", "adapter")
```
Comment (AnastasiiaSergienko) [inline src/main/lua/exasolvs/request_dispatcher.lua:3]: If it's commented out, should it be removed?

File: doc/system_requirements.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
+<head><link href="oft_spec.css" rel="stylesheet"></head>
+
+# System Requirement Specification Exasol Row Level Security
+
+## Introduction
+
+Exasol Row Level Security (short "RLS") is a plug-in for Exasol that provides the ability to grant access on individual rows based on contents of the database.
```
Comment (AnastasiiaSergienko) [inline doc/system_requirements.md:7]:
```suggestion
Exasol Row Level Security (short "RLS") is a plug-in for Exasol that provides the ability to grant access to individual rows based on the contents of the database.
```

File: doc/system_requirements.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+Needs: req
+
+## Functional Requirements
+
+### Row Level Security With Roles
+
+#### Assigning Roles to Users
+`req~assigning-roles-to-users~1`
+
+Data Owners can assign between zero and 63 roles to users.
```
Comment (AnastasiiaSergienko) [inline doc/system_requirements.md:74]:
```suggestion
Data Owners can assign roles between zero and 63 to users.
```

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
I liked the diagrams, I think they are very helpful!
I have a few findings and questions.

File: doc/system_requirements.md (hunk 6 above)
Comment (redcatbear) [reply doc/system_requirements.md:74]: This is meant to be the total amount, not the internal IDs.

File: src/main/lua/entry.lua (hunk 2 above)
Comment (redcatbear) [reply src/main/lua/entry.lua:4]: Was meant to say "creating and wiring up".

File: doc/model/diagrams/activity/act_protection_evaluation.plantuml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+@startuml
+!include ../clean.skin
+!pragma useVerticalIf on
+
+title: Row Protection Evaluation
+
+start
+
+if (row has group\nand row has tenant) then (yes)
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/model/diagrams/activity/act_protection_evaluation.plantuml:17]: Indeed. Fixed.

File: doc/model/diagrams/activity/act_rewrite_with_protection.plantuml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+!include ../clean.skin
+
+title: Query rewriting
+
+start
+
+if (protected) then (yes)
+    group rewrite protected\nselect list
+        if(select in request) then (present)
+            :expand * to all\npayload columns;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/model/diagrams/activity/act_rewrite_with_protection.plantuml:11]: Improved.

File: doc/model/diagrams/activity/act_rewrite_with_protection.plantuml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            :create group\nsub-select filter;
+    
+        case(tenant + role)
+            :create tenant and\nrole-mask\nsub-select filter;
+    
+        case(tenant\n+ group)
+            :crate tenant\nand ngroup\nsub-select filter;
+    
+        endswitch
+        :merge original filter with protection filter;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/model/diagrams/activity/act_rewrite_with_protection.plantuml:36]: Mentioned `FROM` clause.

File: doc/model/diagrams/component/comp_rls_administration_scripts.plantuml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    ' [dsn->req~removing-roles-from-users~1]
+    component delete_rls_role
+
+    component bit_positions 
+
+    ' [dsn->req~listing-all-roles~1]
+    component list_all_roles
+
+    ' [dsn->req~listing-users-and-roles~1]
+    component list_users_and_roles
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/model/diagrams/component/comp_rls_administration_scripts.plantuml:33]: You are absolutely right. Added missing requirement and script.

File: doc/model/diagrams/sequence/seq_entry.plantuml (hunk 1 above)
Comment (redcatbear) [reply doc/model/diagrams/sequence/seq_entry.plantuml:20]: Made more explicit: self-pointer

File: src/main/lua/entry.lua (hunk 2 above)
Comment (redcatbear) [reply src/main/lua/entry.lua:4]: Fixed.

File: src/main/lua/exasolvs/request_dispatcher.lua (hunk 4 above)
Comment (redcatbear) [reply src/main/lua/exasolvs/request_dispatcher.lua:3]: Artifact from before introducing `entry` object. Removed.

File: doc/system_requirements.md [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+This allows data owners to review existing users and their roles.
+
+Covers:
+
+* [feat~row-level-security~1](#row-level-security)
+
+Needs: dsn
+
+#### Listing a User's  Roles
```
Comment (AnastasiiaSergienko) [inline doc/system_requirements.md:153]:
```suggestion
#### Listing a User's Roles
```

File: doc/system_requirements.md [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+`req~listing-user-roles~1`
+
+Data owners can the roles assigned to a single user.
```
Comment (AnastasiiaSergienko) [inline doc/system_requirements.md:156]:
```suggestion
Data owners can list the roles assigned to a single user.
```

**Review by AnastasiiaSergienko — APPROVED** [review]
Approved, but please check my small findings

**Review by AnastasiiaSergienko — APPROVED** [review]
:+1:

**Pull request merged**